### DISCUSSION TITLE
Use builders in gacela.php

### DISCUSSION THREAD

**Discussion by Chemaclass** [gacela-project/gacela#100] [Ideas]
Currently the AbstractConfigGacela's methods returns arrays.
I was thinking about changing the API for the methods to help the user know what can they do with them

### Current
```php
return static fn() => new class () extends AbstractConfigGacela {
    public function config(): array
    {
        return [
            [
                'path' => 'config/*.php',
                'path_local' => 'config/local.php',
                'reader' => PhpConfigReader::class
            ],
            [
                'path' => 'config/.env*',
                'reader' => EnvConfigReader::class
            ],
        ];
    }

    public function mappingInterfaces(): void
    {
        return [
            OneInterface::class => new ConcreteClass(true, 'string', 1, 1.2, ['array']),
            TwoInterface::class => TwoConcrete::class,
        ];
    }
}
```

### Using builder and fluent api

```php
return static fn() => new class () extends AbstractConfigGacela {
    public function config(ConfigBuilder $configBuilder): void
    {
        $configBuilder
            ->addReader(PhpConfigReader::class)
            ->addReader(EnvConfigReader::class)
            ->addPath('config/*.php')
            ->addPath('config/.env*')
            ->addPathLocal('config/local.php');
    }

    public function mappingInterfaces(InterfacesContainer $container): void
    {
        $container->bind(OneInterface::class, new ConcreteClass(true, 'string', 1, 1.2, ['array']));
        $container->bind(TwoInterface::class, TwoConcrete::class);
    }
}
```
This is the draft idea that I wanted to share before starting implementing it. Any thoughts?

**Comment by JesusValeraDev** [comment]
Suggestion to the `config()` method
```php
public function config(ConfigBuilder $configBuilder): void
{
    $configBuilder
        ->addReader(PhpConfigReader::class)
        ->addPath('config/*.php')
        ->addPathLocal('config/local.php')
        ->build();

    $configBuilder
        ->addReader(EnvConfigReader::class)
        ->addPaths(['config/.env*', 'configuration/.env*']) // ??? discussed down 👇🏼
        ->build();
}
```

We are using the `glob()` method, so, I am not sure if could make sense to have more than one path in a reader, but in affirmative case;
1. `$configBuilder->addPath('string1')->addPath('string2')`
2. `$configBuilder->addPaths(['string1', 'string2'])`

Do you think it could worth the `addPaths([])` or maybe multiple calls to `addPath()` is enough?

---

Regarding the public interface of the `$configBuilder` object.
I would suggest to use the `build()` method in order to create `items` independently, this way, we can specify the properties every item will have instead of mixing all configuration in a single `$configBuilder` class as you proposed in the initial example.

What do you think @Chemaclass?

**Reply by Chemaclass** [reply]
Now that the "configReaders" are independent from each config item line, it might make sense to use the `build()` method to separate each config item group independently as you said. Good note 📝 !

**Comment by JesusValeraDev** [comment]
Should we define also an interface for the `overrideResolvableTypes()` method?
Currently is as follows:

```php
public function overrideResolvableTypes(): array
{
    return [
        'Factory' => 'FactCustom',
        'Config' => 'ConfCustom',
        'DependencyProvider' => 'DepProvCustom',
    ];
}
```

What about?
```php
public function overrideResolvableTypes(ResolvableTypes $resolvableTypes): void
{
    $resolvableTypes->setFactoryName('FactCustom')
        ->setConfigName('ConfCustom')
        ->setDependencyProviderName('DepProvCustom');
}
```

**Reply by Chemaclass** [reply]
I like this one a lot!

**Comment by Chemaclass** [comment]
This idea was already fully implemented here https://github.com/gacela-project/gacela/pull/146 embracing an object `GacelaConfig` instead of a key-value array. You can see a complete example of gacela.php here: https://gacela-project.com/docs/bootstrap/#a-complete-example-using-gacela-php
